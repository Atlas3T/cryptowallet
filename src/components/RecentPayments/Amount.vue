<!--

it renders a div with amount-ox class
it variant grey is passed, it renders grey-text class
it rendersp properly formatted amount

 -->


<template>
  <div
    :class="{ 'grey-text': variant.indexOf('grey') != -1 }"
    class="amount-box"
  >
    <div>{{ formattedAmount }}</div>
  </div>
</template>

<script>
import numeral from 'numeral';

export default {
  name: 'Amount',
  props: {
    amount: {
      type: Number,
      required: true,
    },
    variant: {
      type: String,
      default() {
        return '';
      },
    },
  },
  computed: {
    formattedAmount() {
      const formatted = `${numeral(this.amount).format('0.00')} Â£`;
      return (this.amount > 0 ? `+${formatted}` : formatted);
    },
  },
};
</script>

<style scoped>
  .amount-box {
    display: flex;
    justify-content: flex-end;
    text-align: right;
    white-space: nowrap;
  }
</style>
