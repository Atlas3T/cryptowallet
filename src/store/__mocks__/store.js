import Vue from 'vue';
import Vuex from 'vuex';
import VuexORM from '@vuex-orm/core';
import VuexORMLoki from '@atlascity/vuex-orm-plugin-lokijs';

import wallet from '@/store/wallet';
import contacts from '@/store/contacts/__mocks__/contacts.js';
import search from '@/store/search/__mocks__/search.js';
import settings from '@/store/settings/__mocks__/settings.js';
import setup from '@/store/setup/__mocks__/setup.js';
import qrcode from '@/store/qrcode/__mocks__/qrcode.js';
import modals from '@/store/modals/__mocks__/modals.js';

// entities
import Account from '@/store/wallet/entities/account';
import Address from '@/store/wallet/entities/address';
import Tx from '@/store/wallet/entities/tx';
import Utxo from '@/store/wallet/entities/utxo';
import Wallet from '@/store/wallet/entities/wallet';
import Coin from '@/store/wallet/entities/coin';
import KeyPair from '@/store/wallet/entities/keyPair';
import Prices from '@/store/prices';
import LatestPrice from '@/store/latestPrice';

Vue.use(Vuex);

// Setup ORM database.
const database = new VuexORM.Database();

database.register(Account, {});
database.register(Address, {});
database.register(Tx, {});
database.register(Utxo, {});
database.register(Wallet, {});
database.register(Coin, {});
database.register(KeyPair, {});
database.register(Prices, {});
database.register(LatestPrice, {});

Vue.prototype.encryptedModels = [
  Account,
  Wallet,
];

const options = {
  env: 'browser',
};

VuexORM.use(VuexORMLoki, { database, options });

export const actions = {};
export const getters = {};
export const mutations = {};
export const state = {};

export function createMocks(custom = {
  getters: {}, mutations: {}, actions: {}, state: {},
}) {
  /**
   * mock store modules.
   */
  const modules = {
    wallet,
    search,
    settings,
    contacts,
    setup,
    qrcode,
    modals,
  };

  let mockGettersCombined = {};
  let mockMutationsCombined = {};
  let mockActionsCombined = {};
  let mockStateCombined = {};

  Object.entries(modules).forEach(([name, module]) => {
    mockGettersCombined = Object.assign({}, mockGettersCombined, module.getters);
    mockMutationsCombined = Object.assign({}, mockMutationsCombined, module.mutations);
    mockActionsCombined = Object.assign({}, mockActionsCombined, module.actions);
    mockStateCombined = Object.assign({}, mockStateCombined, module.state);

    if (custom.state && custom.state[name]) {
      module.state = Object.assign(module.state, custom.state[name]);
    }

    modules[name] = {
      namespaced: true,
      getters: Object.assign({}, getters, module.getters),
      mutations: Object.assign({}, mutations, module.mutations),
      actions: Object.assign({}, actions, module.actions),
      state: Object.assign({}, state, module.state),
    };
  });

  mockGettersCombined = Object.assign({}, mockGettersCombined, custom.getters);
  mockMutationsCombined = Object.assign({}, mockMutationsCombined, custom.mutations);
  mockActionsCombined = Object.assign({}, mockActionsCombined, custom.actions);
  mockStateCombined = Object.assign({}, mockStateCombined, custom.state);
  return {
    getters: mockGettersCombined,
    mutations: mockMutationsCombined,
    actions: mockActionsCombined,
    state: mockStateCombined,
    store: new Vuex.Store({
      getters: mockGettersCombined,
      mutations: mockMutationsCombined,
      actions: mockActionsCombined,
      state: mockStateCombined,
      modules,
      plugins: [VuexORM.install(database)],
    }),
  };
}

export const { store } = createMocks();
