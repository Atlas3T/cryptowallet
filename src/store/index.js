/*eslint-disable*/

import Vue from 'vue';
import Vuex from 'vuex';
import VuexPersistence from 'vuex-persist';
import VuexORM from '@vuex-orm/core';
import VuexORMLoki from 'vuex-orm-lokijs';

// import entities.
import Account from './wallet/entities/account';
import Address from './wallet/entities/address';
import Tx from './wallet/entities/tx';
import Utxo from './wallet/entities/utxo';
import Wallet from './wallet/entities/wallet';
import Coin from './wallet/entities/coin';
import KeyPair from './wallet/entities/keyPair';
import Prices from './prices';
import LatestPrice from './latestPrice';

// import modules.
import payments from './payments';
import search from './search';
import settings from './settings';
import setup from './setup';
import contacts from './contacts';

Vue.use(Vuex);

// Setup ORM database.
const database = new VuexORM.Database();

database.register(Account, {});
database.register(Address, {});
database.register(Tx, {});
database.register(Utxo, {});
database.register(Wallet, {});
database.register(Coin, {});
database.register(KeyPair, {});
database.register(Prices, {});
database.register(LatestPrice, {});

Vue.prototype.encryptedModels = [
  Account,
];

window.Tx = Tx;
window.Utxo = Utxo;
window.Address = Address;
window.Wallet = Wallet;
window.LatestPrice = LatestPrice;
window.Prices = Prices;
window.Account = Account;
window.Coin = Coin;

// Setup persistant storage.
const vuexLocal = new VuexPersistence({
  supportCircular: true,
  storage: window.localStorage,
  modules: [settings],
});


const options = {
  env: 'browser',
};

function hydrationCompletedCallback() {
  setTimeout(() => {
    store.dispatch('settings/setLoading', false);
  }, 1000);
}

VuexORM.use(VuexORMLoki, { database, options, hydrationCompletedCallback });


/**
 * Create CryptoWallet Vuex store obj.
 * @type {Store<any>}
 */
const store = new Vuex.Store({
  modules: {
    payments,
    search,
    settings,
    setup,
    contacts,
  },
  plugins: [VuexORM.install(database)],
});

Vue.prototype.activeWallets = {};

// @todo remove line below (used for development purposes only)
window.store = store;

export default store;
