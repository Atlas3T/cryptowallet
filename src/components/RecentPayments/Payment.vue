<template>
  <div class="columns is-mobile no-margin-bottom">
    <div class="column is-2">
      <div class="circle"/>
    </div>
    <div class="column">
      <div class="columns is-mobile no-margin-bottom">
        <div class="column is-7 no-padding-bottom">{{ payment.title }}</div>
        <div class="column no-padding-bottom">
          <Amount
            :amount="payment.amount"
            variant="['grey']"
          />
        </div>
      </div>
      <div class="grey-text">
        {{ paymentTime }}
      </div>
    </div>
  </div>
</template>

<script>
import Amount from './Amount.vue';

export default {
  name: 'Payment',
  components: {
    Amount,
  },
  props: {
    payment: {
      type: Object,
      required: true,
    },
  },
  computed: {
    /*
     * converts timestamp into full date string and then extracts only the time out of it (HH:mm)
     * @return {String} eg. "02:59"
     */
    paymentTime() {
      return new Date(this.payment.ts).toTimeString().split(' ')[0].slice(0, 5);
    },
  },
};
</script>

<style scoped>
.no-padding-bottom {
  padding-bottom: 0;
}

.no-margin-bottom {
  margin-bottom: 0;
}

.circle {
  background: #ececec;
  border-radius: 100%;
  height: 2.5rem;
  width: 2.5rem;
  margin-top: 5px;
}
</style>
