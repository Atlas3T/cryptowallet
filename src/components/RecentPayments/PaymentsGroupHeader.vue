<template>
  <div class="columns is-mobile payments-group-header">
    <div class="column">
      {{ group.date }}
    </div>
    <div class="column">
      <Amount :amount="balanceChange"/>
    </div>
  </div>
</template>

<script>
import Amount from './Amount.vue';

export default {
  name: 'PaymentsGroupHeader',
  components: {
    Amount,
  },
  props: {
    group: {
      type: Object,
      required: true,
    },
  },
  computed: {
    balanceChange() {
      return this.group.payments.reduce((accumulator, payment) => accumulator + payment.amount, 0);
    },
  },
};
</script>

<style scoped>
.payments-group-header {
  margin-bottom: -1rem;
  margin-top: 0.5rem;
}
</style>
