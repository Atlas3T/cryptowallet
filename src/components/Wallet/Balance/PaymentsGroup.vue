<template>
  <section>
    <PaymentsGroupHeader :group="group" />
    <Payment
      v-for="payment in group.payments"
      :key="payment.ts"
      :payment="payment"
    />
  </section>
</template>

<script>
import PaymentsGroupHeader from './PaymentsGroupHeader.vue';
import Payment from './Payment.vue';

export default {
  name: 'PaymentsGroup',
  components: {
    PaymentsGroupHeader,
    Payment,
  },
  props: {
    group: {
      type: Object,
      required: true,
    },
  },
};
</script>
