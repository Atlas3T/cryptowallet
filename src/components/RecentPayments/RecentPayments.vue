<template>
  <section class="recent-payments">
    <PaymentsGroup
      v-for="group in groupedPayments"
      :key="group"
      :group="group"
    />
  </section>
</template>

<script>
import PaymentsGroup from './PaymentsGroup.vue';

export default {
  name: 'RecentPayments',
  components: {
    PaymentsGroup,
  },

  computed: {
    payments() {
      return this.$store.state.wallet.payments;
    },
    groupedPayments() {
      const sortedByDate = this.payments.slice().sort((a, b) => b.ts - a.ts);
      const groupedPayments = [];

      sortedByDate.forEach((payment) => {
        let timeStamp = new Date(payment.ts * 1000);
        timeStamp = `${timeStamp.getDay} ${timeStamp.getYear}`;
        const groupItem = groupedPayments.find(item => item.date === timeStamp);

        if (!groupItem) {
          groupedPayments.push({
            date: timeStamp,
            payments: [],
          });
        }

        groupedPayments.map((item) => {
          if (item.date === timeStamp) item.payments.push(payment);
          return item;
        });
      });

      return groupedPayments;
    },
  },
};
</script>

<style scoped>
  .recent-payments {
    padding: 0rem 1rem 1rem 1rem;
    border-top: 1px solid grey;
    border-bottom: 1px solid grey;
    position: relative;
    overflow-y: scroll;
    margin: 0 10px;
  }
</style>
