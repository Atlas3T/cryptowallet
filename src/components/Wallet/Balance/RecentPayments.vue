<template>
  <section class="recent-payments">
    <q-scroll-area class="recent-payments-scroll-area">
      <div class="scroll-content">
        <PaymentsGroup
          v-for="group in groupedPayments"
          :key="group.key"
          :group="group"
        />
      </div>
    </q-scroll-area>
  </section>
</template>
<script>
import PaymentsGroup from './PaymentsGroup.vue';

const monthNames = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];

export default {
  name: 'RecentPayments',
  components: {
    PaymentsGroup,
  },
  computed: {
    payments() {
      return this.$store.state.payments.payments;
    },
    /*
     * raw payments data is a flat array, it has to be transformed first,
     * payments have to be grouped by date, the function returns a groupedPayments Array
     * each group in the Array represents a day, eg. 12 November
     * each group will include payments that happened on that day
     * @return {Array}
     */
    groupedPayments() {
      const sortedByDate = this.payments.slice().sort((a, b) => b.ts - a.ts);
      const groupedPayments = [];
      let groupKey = 0;

      sortedByDate.forEach((payment) => {
        const day = new Date(payment.ts).getDate();
        const month = monthNames[new Date(payment.ts).getMonth()];
        const paymentDate = `${day} ${month}`;
        const groupItem = groupedPayments.find(item => item.date === paymentDate);

        if (!groupItem) {
          groupedPayments.push({
            key: groupKey,
            date: paymentDate,
            payments: [],
          });

          groupKey += 1;
        }

        groupedPayments.map((item) => {
          if (item.date === paymentDate) item.payments.push(payment);
          return item;
        });
      });

      return groupedPayments;
    },
  },
};
</script>

<style scoped>
  .recent-payments-scroll-area {
    width: auto;
    height: 100%;
  }

  .recent-payments {
    height: 60vh;
    border-top: 2px solid #09233a;
    position: relative;
    background: #16324a;
  }

  .scroll-content {
    overflow: hidden;
    padding: 0 1rem;
  }
</style>
