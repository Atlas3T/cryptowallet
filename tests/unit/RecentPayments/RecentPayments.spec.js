import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import RecentPayments from '@/components/RecentPayments/RecentPayments.vue';
import PaymentsGroup from '@/components/RecentPayments/PaymentsGroup.vue';
import { __createMocks as createStoreMocks } from '@/../tests/store/__mocks__/store.js';

const localVue = createLocalVue();
localVue.use(Vuex);

const monthNames = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];

describe('RecentPayments.vue', () => {
  let storeMocks;
  let wrapper;
  let store;

  function wrapperInit (options) {
    return shallowMount(RecentPayments, options);
  }

  function storeInit (custom) {
    storeMocks = createStoreMocks(custom);
    wrapper = wrapperInit({ localVue, store: storeMocks.store });
    store = wrapper.vm.$store;
  }

  beforeEach(() => storeInit());

  it('renders and matches snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders a section with .recent-payments class', () => {
    expect(wrapper.contains('section.recent-payments')).toBe(true);
  });

  it('loops through the data and renders multiple PaymentsGroup components', () => {
    expect(wrapper.findAll(PaymentsGroup).length).toBeGreaterThan(1);
  });

  it('it computes groupedPayments correctly', () => {
    expect(wrapper.vm.groupedPayments.length).toBe(3);

    wrapper.vm.groupedPayments.forEach((group) => {
      expect(group.payments.length > 0).toBe(true);

      group.payments.forEach((payment) => {
        const day = new Date(payment.ts).getDate();
        const month = monthNames[new Date(payment.ts).getMonth()];
        const paymentDate = `${day} ${month}`;
        expect(paymentDate === group.date).toBe(true);
      });
    });
  });
});
