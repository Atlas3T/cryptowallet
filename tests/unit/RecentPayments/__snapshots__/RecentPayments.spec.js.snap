// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecentPayments.vue renders and matches snapshot 1`] = `
<section
  class="recent-payments"
>
  <paymentsgroup-stub
    group="[object Object]"
  />
  <paymentsgroup-stub
    group="[object Object]"
  />
  <paymentsgroup-stub
    group="[object Object]"
  />
</section>
`;
